File: client-flutter/lib/widgets/invoice_detail_sheet.dart
import 'dart:typed_data';
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';
import '../config/app_colors.dart';
import '../models/invoice.dart';
import '../services/invoice_service.dart';

/// Shared invoice detail bottom sheet with QR-Rechnung support.
///
/// Usage:
/// ```dart
/// InvoiceDetailSheet.show(context, invoice: myInvoice);
/// ```
class InvoiceDetailSheet extends StatefulWidget {
  final Invoice invoice;

  const InvoiceDetailSheet({
    super.key,
    required this.invoice,
  });

  /// Convenience: opens the sheet as a modal bottom sheet.
  static void show(BuildContext context, {
    required Invoice invoice,
  }) {
    showModalBottomSheet(
      context: context,
      backgroundColor: AppColors.surface,
      isScrollControlled: true,
      shape: const RoundedRectangleBorder(
        borderRadius: BorderRadius.vertical(top: Radius.circular(20)),
      ),
      builder: (_) => InvoiceDetailSheet(
        invoice: invoice,
      ),
    );
  }

  @override
  State<InvoiceDetailSheet> createState() => _InvoiceDetailSheetState();
}

class _InvoiceDetailSheetState extends State<InvoiceDetailSheet> {
  final InvoiceService _service = InvoiceService();
  bool _qrLoading = false;
  Uint8List? _qrImage;
  String? _qrError;

  Invoice get inv => widget.invoice;

  String _fmt(DateTime? dt) =>
      dt == null ? '-' : DateFormat('dd.MM.yyyy').format(dt);

  Future<void> _loadQr() async {
    setState(() { _qrLoading = true; _qrError = null; });
    final bytes = await _service.fetchQrBill(
      invoiceNumber: inv.invoiceNumber,
    );
    if (!mounted) return;
    setState(() {
      _qrLoading = false;
      if (bytes != null) {
        _qrImage = bytes;
      } else {
        _qrError = 'QR-Rechnung konnte nicht geladen werden.';
      }
    });
  }

  @override
  Widget build(BuildContext context) {
    final isPaid = inv.isPaid;
    final statusColor = isPaid ? AppColors.green : AppColors.primary;
    final statusBg = isPaid
        ? AppColors.green.withOpacity(0.12)
        : AppColors.primary.withOpacity(0.12);

    return DraggableScrollableSheet(
      expand: false,
      initialChildSize: 0.55,
      minChildSize: 0.35,
      maxChildSize: 0.92,
      builder: (context, scrollController) => SingleChildScrollView(
        controller: scrollController,
        padding: const EdgeInsets.fromLTRB(24, 12, 24, 32),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: [
            // ── Drag handle ──
            Container(
              width: 40, height: 4,
              decoration: BoxDecoration(
                color: AppColors.muted.withOpacity(0.3),
                borderRadius: BorderRadius.circular(2),
              ),
            ),
            const SizedBox(height: 20),

            // ── Header row ──
            Row(
              children: [
                Container(
                  width: 48, height: 48,
                  decoration: BoxDecoration(
                    color: AppColors.primary.withOpacity(0.12),
                    borderRadius: BorderRadius.circular(12),
                  ),
                  child: const Icon(Icons.receipt_rounded,
                      color: AppColors.primary, size: 24),
                ),
                const SizedBox(width: 14),
                Expanded(
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text('Rechnung ${inv.invoiceNumber}',
                          style: const TextStyle(
                              color: AppColors.text,
                              fontSize: 18,
                              fontWeight: FontWeight.w800)),
                      const SizedBox(height: 2),
                      Container(
                        padding: const EdgeInsets.symmetric(
                            horizontal: 8, vertical: 3),
                        decoration: BoxDecoration(
                            color: statusBg,
                            borderRadius: BorderRadius.circular(6)),
                        child: Text(isPaid ? 'Bezahlt' : 'Offen',
                            style: TextStyle(
                                color: statusColor,
                                fontSize: 11,
                                fontWeight: FontWeight.w700)),
                      ),
                    ],
                  ),
                ),
                Text(
                    '${inv.currency} ${inv.amount.toStringAsFixed(2)}',
                    style: const TextStyle(
                        color: AppColors.text,
                        fontSize: 20,
                        fontWeight: FontWeight.w800)),
              ],
            ),
            const SizedBox(height: 20),
            const Divider(color: AppColors.border, height: 1),
            const SizedBox(height: 16),

            // ── Detail rows ──
            if (inv.packageName != null)
              _detailRow('Paket', inv.packageName!),
            if (inv.credits != null)
              _detailRow('Lektionen',
                  '${inv.credits} × 60 Min. Personal Training'),
            if (inv.durationMonths != null)
              _detailRow('Gültigkeit',
                  '${inv.durationMonths} ${inv.durationMonths == 1 ? "Monat" : "Monate"}'),
            _detailRow('Rechnungsdatum', _fmt(inv.transactionDate)),
            _detailRow('Fällig am', _fmt(inv.dueDate)),
            const SizedBox(height: 16),

            // ── Payment info box ──
            Container(
              width: double.infinity,
              padding: const EdgeInsets.all(14),
              decoration: BoxDecoration(
                color: AppColors.primary.withOpacity(0.06),
                borderRadius: BorderRadius.circular(10),
                border:
                    Border.all(color: AppColors.primary.withOpacity(0.15)),
              ),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  const Text('Zahlungsinformationen',
                      style: TextStyle(
                          color: AppColors.text,
                          fontSize: 13,
                          fontWeight: FontWeight.w700)),
                  const SizedBox(height: 6),
                  Text(
                      'SIHLHEALTH GmbH\nVerwendungszweck: ${inv.invoiceNumber}',
                      style: const TextStyle(
                          color: AppColors.muted,
                          fontSize: 12,
                          height: 1.5)),
                ],
              ),
            ),

            // ── QR-Rechnung section ──
            if (!isPaid) ...[
              const SizedBox(height: 16),
              if (_qrImage == null && !_qrLoading && _qrError == null)
                SizedBox(
                  width: double.infinity,
                  child: OutlinedButton.icon(
                    onPressed: _loadQr,
                    icon: const Icon(Icons.qr_code_rounded, size: 18),
                    label: const Text('QR-Rechnung anzeigen'),
                    style: OutlinedButton.styleFrom(
                      foregroundColor: AppColors.primary,
                      side: BorderSide(
                          color: AppColors.primary.withOpacity(0.4)),
                      padding: const EdgeInsets.symmetric(vertical: 12),
                      shape: RoundedRectangleBorder(
                          borderRadius: BorderRadius.circular(10)),
                    ),
                  ),
                ),
              if (_qrLoading)
                const Padding(
                  padding: EdgeInsets.symmetric(vertical: 20),
                  child: SizedBox(
                    width: 28, height: 28,
                    child: CircularProgressIndicator(
                      strokeWidth: 2.5,
                      color: AppColors.primary,
                    ),
                  ),
                ),
              if (_qrError != null) ...[
                Padding(
                  padding: const EdgeInsets.only(top: 8),
                  child: Text(_qrError!,
                      style: const TextStyle(
                          color: AppColors.muted, fontSize: 12)),
                ),
                const SizedBox(height: 8),
                TextButton.icon(
                  onPressed: _loadQr,
                  icon: const Icon(Icons.refresh_rounded, size: 16),
                  label: const Text('Erneut versuchen'),
                  style: TextButton.styleFrom(foregroundColor: AppColors.primary),
                ),
              ],
              if (_qrImage != null) ...[
                const SizedBox(height: 12),
                Container(
                  decoration: BoxDecoration(
                    color: Colors.white,
                    borderRadius: BorderRadius.circular(10),
                    border: Border.all(color: AppColors.border),
                  ),
                  clipBehavior: Clip.antiAlias,
                  child: Image.memory(
                    _qrImage!,
                    fit: BoxFit.contain,
                    width: double.infinity,
                  ),
                ),
              ],
            ],
          ],
        ),
      ),
    );
  }

  static Widget _detailRow(String label, String value) {
    return Padding(
      padding: const EdgeInsets.only(bottom: 10),
      child: Row(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          SizedBox(
            width: 130,
            child: Text(label,
                style:
                    const TextStyle(color: AppColors.muted, fontSize: 13)),
          ),
          Expanded(
            child: Text(value,
                style: const TextStyle(
                    color: AppColors.text,
                    fontSize: 13,
                    fontWeight: FontWeight.w600)),
          ),
        ],
      ),
    );
  }
}

File: client-ios/SihlClient/Views/Profile/InvoiceDetailSheet.swift
import SwiftUI

/// Pendant zu `widgets/invoice_detail_sheet.dart`.
/// Wird als `.sheet(item:)` aus ProfileView geöffnet.
struct InvoiceDetailSheet: View {
    let invoice: Invoice
    @Environment(\.dismiss) private var dismiss

    @State private var qrData: Data?
    @State private var qrLoading = false
    @State private var qrError: String?

    private let svc = InvoiceService()

    private static let dateFmt: DateFormatter = {
        let f = DateFormatter(); f.dateFormat = "dd.MM.yyyy"; return f
    }()
    private func fmt(_ date: Date?) -> String {
        date.map { Self.dateFmt.string(from: $0) } ?? "-"
    }

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: 0) {
                    header
                        .padding(.bottom, 20)

                    Divider().background(AppColor.border)
                        .padding(.bottom, 16)

                    detailRows
                        .padding(.bottom, 16)

                    paymentInfoBox
                        .padding(.bottom, invoice.isPaid ? 0 : 16)

                    if !invoice.isPaid {
                        qrBillSection
                    }
                }
                .padding(.horizontal, AppSpacing.screen)
                .padding(.vertical, 20)
            }
            .background(AppColor.background.ignoresSafeArea())
            .navigationTitle("Rechnung \(invoice.invoiceNumber)")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .topBarTrailing) {
                    Button("Fertig") { dismiss() }
                        .foregroundStyle(AppColor.primary)
                }
            }
        }
    }

    // MARK: - Header

    private var header: some View {
        HStack(spacing: 14) {
            // Icon
            RoundedRectangle(cornerRadius: AppRadius.control)
                .fill(AppColor.primary.opacity(0.12))
                .frame(width: 48, height: 48)
                .overlay(Image(systemName: "receipt")
                    .font(.app(20))
                    .foregroundStyle(AppColor.primary))

            // Number + Status
            VStack(alignment: .leading, spacing: 4) {
                Text("Rechnung \(invoice.invoiceNumber)")
                    .font(.app(18, weight: .heavy))
                    .foregroundStyle(AppColor.text)
                statusBadge
            }

            Spacer()

            // Amount
            Text("\(invoice.currency) \(String(format: "%.2f", invoice.amount))")
                .font(.app(20, weight: .heavy))
                .foregroundStyle(AppColor.text)
        }
    }

    private var statusBadge: some View {
        let color = invoice.isPaid ? AppColor.green : AppColor.primary
        return Text(invoice.isPaid ? "Bezahlt" : "Offen")
            .font(.app(11, weight: .bold))
            .foregroundStyle(color)
            .padding(.horizontal, 8)
            .padding(.vertical, 3)
            .background(color.opacity(0.12))
            .clipShape(RoundedRectangle(cornerRadius: 6))
    }

    // MARK: - Detail Rows

    private var detailRows: some View {
        VStack(alignment: .leading, spacing: 10) {
            if let pkg = invoice.packageName {
                DetailRow(label: "Paket", value: pkg)
            }
            if let c = invoice.credits {
                DetailRow(label: "Lektionen",
                          value: "\(c) × 60 Min. Personal Training")
            }
            if let m = invoice.durationMonths {
                DetailRow(label: "Gültigkeit",
                          value: "\(m) \(m == 1 ? "Monat" : "Monate")")
            }
            DetailRow(label: "Rechnungsdatum", value: fmt(invoice.transactionDate))
            DetailRow(label: "Fällig am",       value: fmt(invoice.dueDate))
        }
    }

    // MARK: - Payment Info Box

    private var paymentInfoBox: some View {
        VStack(alignment: .leading, spacing: 6) {
            Text("Zahlungsinformationen")
                .font(.app(13, weight: .bold))
                .foregroundStyle(AppColor.text)
            Text("SIHLHEALTH GmbH\nVerwendungszweck: \(invoice.invoiceNumber)")
                .font(.app(12))
                .foregroundStyle(AppColor.muted)
                .lineSpacing(4)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(14)
        .background(AppColor.primary.opacity(0.06))
        .clipShape(RoundedRectangle(cornerRadius: AppRadius.control))
        .overlay(RoundedRectangle(cornerRadius: AppRadius.control)
            .stroke(AppColor.primary.opacity(0.15), lineWidth: 1))
    }

    // MARK: - QR Bill

    private var qrBillSection: some View {
        VStack(spacing: 12) {
            if qrData == nil && !qrLoading && qrError == nil {
                // "QR-Rechnung anzeigen" Button
                Button {
                    Task { await loadQr() }
                } label: {
                    Label("QR-Rechnung anzeigen", systemImage: "qrcode")
                        .font(.app(15, weight: .medium))
                        .foregroundStyle(AppColor.primary)
                        .frame(maxWidth: .infinity)
                        .padding(.vertical, 12)
                        .overlay(RoundedRectangle(cornerRadius: AppRadius.control)
                            .stroke(AppColor.primary.opacity(0.4), lineWidth: 1))
                }
            }

            if qrLoading {
                ProgressView().tint(AppColor.primary).padding(.vertical, 20)
            }

            if let err = qrError {
                VStack(spacing: 8) {
                    Text(err)
                        .font(.app(12))
                        .foregroundStyle(AppColor.muted)
                    Button {
                        Task { await loadQr() }
                    } label: {
                        Label("Erneut versuchen", systemImage: "arrow.clockwise")
                            .font(.app(14))
                            .foregroundStyle(AppColor.primary)
                    }
                }
            }

            if let imgData = qrData, let uiImg = UIImage(data: imgData) {
                Image(uiImage: uiImg)
                    .resizable()
                    .scaledToFit()
                    .frame(maxWidth: .infinity)
                    .background(Color.white)
                    .clipShape(RoundedRectangle(cornerRadius: AppRadius.control))
                    .overlay(RoundedRectangle(cornerRadius: AppRadius.control)
                        .stroke(AppColor.border, lineWidth: 1))
            }
        }
    }

    private func loadQr() async {
        qrLoading = true
        qrError   = nil
        do {
            if let d = try await svc.fetchQrBill(invoiceNumber: invoice.invoiceNumber) {
                qrData = d
            } else {
                qrError = "QR-Rechnung konnte nicht geladen werden."
            }
        } catch {
            qrError = "QR-Rechnung konnte nicht geladen werden."
        }
        qrLoading = false
    }
}

// MARK: - Detail Row (shared)

struct DetailRow: View {
    let label: String
    let value: String

    var body: some View {
        HStack(alignment: .top, spacing: 0) {
            Text(label)
                .font(.app(13))
                .foregroundStyle(AppColor.muted)
                .frame(width: 130, alignment: .leading)
            Text(value)
                .font(.app(13, weight: .semibold))
                .foregroundStyle(AppColor.text)
                .frame(maxWidth: .infinity, alignment: .leading)
        }
    }
}
